### PR TITLE
Rename `submission.dac` to `submission.rems`

### LINKED ISSUE #588: Unify DAC and policy implementations
#### Proposed solution

Currently we have two distinct implementations for DAC and policy.
1. From EGA schema
2. From REMS

and they are not connected in any way.

#### Tasks
- [ ] Design unification for both implementations
The proposed solution would be to:
  1. Rename, in the submission schema `dac` to `rems` https://github.com/CSCfi/metadata-submitter/pull/648
  2. Accept EGA XML and validate against data that lives in REMS - adding EGA DAC object to DB, and configuring the `submission.rems` field
  3. When setting REMS fields through the API (including front-end), create (or replace) the EGA DAC object to DB. #532

- [ ] Make code changes
- [ ] Update tests
- [ ] Create ticket for the frontend

#### DoD (Definition of Done)

DAC and policy implementations are unified, and interchangeable.

### PR DESCRIPTION
**Pull request by csc-jm**
### Description

Separated this task from the other tasks given in #588 for less chaotic reviewing. So this just refactors the naming scheme of the `dac` field in the `submission` object to `rems`.

### Related issues

Partially fixes #588 

### Type of change

- [x] Breaking change (probably for front-end, considering the endpoint name is altered)

### Changes Made

- change the `submission` schema
- change the PUT endpoint name to `/v1/submission/rems`
- modify existing integration tests
- edit openapi specs

### Testing

- [x] Integration Tests

### CODE REVIEW

File: CHANGELOG.md
Comment (blankdots) [inline CHANGELOG.md:21]: maybe mention #648 where it was changed

File: docs/specification.yml
Comment (blankdots) [inline docs/specification.yml:2845]: this is a workflow in REMS not a dac, maybe we can keep terminology, because from REMS workflow we derive a the  DAC metadata specific to the EGA model

**Review by blankdots — COMMENTED** [review]
looks good to me, but i think we should keep terminology consistent with the source, REMS has workflow and approvers/Handlers  make up a DAC

File: docs/specification.yml
Comment (csc-felipe) [reply docs/specification.yml:2845]: It is called `REMS DAC` in the submission schema. Perhaps a new issue for that?

**Review by csc-felipe — COMMENTED** [review]
Looks good to me. Well done, finding all the occurrences to be renamed.

After Stefan's comments are resolved, please dismiss the review and request it again. ^^

File: docs/specification.yml
Comment (csc-jm) [reply docs/specification.yml:2845]: Mmm, I was also wondering what constitutes as "DAC" and just "workflow". As @csc-felipe mentioned, `submission.json` originally titled these as `"REMS DAC for datasets"` but perhaps this was inaccurate titling? Should all of these just be `REMS workflows` instead?

File: CHANGELOG.md
Comment (csc-jm) [reply CHANGELOG.md:21]: Good point

File: docs/specification.yml
Comment (blankdots) [reply docs/specification.yml:2845]: new issue sounds fine.

**Comment by csc-jm** [comment]
While converting all these DAC namings, I also [looked at this](https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/api/handlers/publish.py#L421-L428) but wasn't sure if that needs to change or stay like that.

**Comment by blankdots** [comment]
> While converting all these DAC namings, I also [looked at this](https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/api/handlers/publish.py#L421-L428) but wasn't sure if that dataset key needs to change or stay like that.

:thinking: what is the dataset key in the lines you highlighted, could we get more details ?

**Comment by csc-felipe** [comment]
That's adding the REMS details to a dataset. For each dataset we create a resource, and a catalogue item in REMS. We don't use that information anywhere at the moment, but still, it should be preserved. What line 422 does is that it adds the REMS link that can be used to apply for access to the dataset to the dataset description that will be visible in Etsin.

My answer to @csc-jm would be that `dac` could be renamed to `rems` also in that case. I expect that later the frontend will display some information derived from these values.

**Comment by csc-felipe** [comment]
I do think that writing **REMS DAC** makes sense, as what we're doing is translating REMS data into DAC. But I don't mind if you prefer to have different wording. The most important imo is that it's consistent, and at least somewhat accurate.

**Comment by csc-felipe** [comment]
On the other hand, writing some documentation how the publishing process works would be good. I just wonder if now is a good time, or after we make the changes we've been planning.

**Comment by blankdots** [comment]
> That's adding the REMS details to a `dataset`. For each `dataset` we create a `resource`, and a `catalogue item` in REMS. We don't use that information anywhere at the moment, but still, it should be preserved. What line 422 does is that it adds the REMS link that can be used to apply for access to the `dataset` to the `dataset description` that will be visible in Etsin.


hmm i see, so it is added to these object this info: https://github.com/CSCfi/metadata-submitter/blob/develop/metadata_backend/helpers/schemas/ena_dataset.json and https://github.com/CSCfi/metadata-submitter/blob/develop/metadata_backend/helpers/schemas/ena_bpdataset.json - but we also have this information in https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/helpers/schemas/submission.json#L1165 is that correct ?

My question would be what do we do with a submission that has multiple datasets each with their own worflow in REMS because a submission can have multiple datasets and for each one in rems you need to create a separate catalogue item, however https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/helpers/schemas/submission.json#L1165 allows for only one object not an array ?

**Review by blankdots — CHANGES_REQUESTED** [review]
> On the other hand, writing some documentation how the publishing process works would be good. I just wonder if now is a good time, or after we make the changes we've been planning.

yeah, i think this needs to come with some documentation on what represents a REMS DAC and how we map concepts, otherwise this will not go well with further development

**Comment by csc-felipe** [comment]
> but we also have this information in

@blankdots, not really. Most of the information should be the same, but there are values which are specific to the dataset: `resource_id` and `catalogue_item_id`. Also the URL which is build from those values.

**Comment by csc-felipe** [comment]
As for the second question, in a meeting with other people, we decided that a submission has the same workflow / policy / dac for all datasets. That's why it's an object in the submission.

**Comment by blankdots** [comment]
> > but we also have this information in
> 
> @blankdots, not really. Most of the information should be the same, but there are values which are specific to the dataset: `resource_id` and `catalogue_item_id`. Also the URL which is build from those values.

then we need to remove these lines:
- https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/helpers/schemas/submission.json#L1190
- https://github.com/CSCfi/metadata-submitter/blob/refactor/rename-submission-dac/metadata_backend/helpers/schemas/submission.json#L1194

**Comment by csc-felipe** [comment]
Indeed, there was a bug. The issue was created :) https://github.com/CSCfi/metadata-submitter/issues/649

**Comment by csc-jm** [comment]
Moved this momentarily to WIP because I need to push and test some things. Sorry if it still generated bunch unnecessary emails. I renamed the `dac` field of a dataset. I almost also fixed the issue #649 opened by @csc-felipe in this PR but wanted to clarify something first so didn't. Before this PR gets more confusing to review, we can maybe merge and follow up with more issues

**Review by blankdots — APPROVED** [review]
i am ok with changes so far.

**Pull request merged**
